File: programmers/lv2/전화번호 목록.py
"""
    lv.2 전화번호 목록
    
    1.  전화번호부의 전화번호들을 딕서너리에 우선 등록.
    2.  전화번호 하나하나들을 앞에서부터 분해해 만들 수 있는 접두사 조합들에 대해,
        전화번호와 동일하지 않은 접두사가 존재할 경우 참.
"""


def solution(phone_book):
    phoneDict = {}

    for pn in phone_book:
        phoneDict[pn] = True

    for pn in phone_book:
        partial = ""

        for c in pn:
            partial += c

            if partial in phoneDict and partial != pn:
                return False

    return True

File: programmers/lv2/전화번호 목록.js
/* 
  lv.2 전화번호 목록 

  1.  주어진 배열을 순회하면서 접두사가 존재하는 지 확인하는 문제.
  2.  백만 길이의 배열이라도 정렬 과정을 거친 후 검사하는 것이 확실히 빠름.
*/

function solution(phone_book) {
  var answer = true;

  answer = phone_book
    .sort()
    .every((num, idx, arr) => !arr[idx + 1]?.startsWith(num));

  return answer;
}

const phone_book = ["123", "456", "789"];

console.log(solution(phone_book));
